### PR TITLE
[powermax] Introduce some new channels for better status reporting

### LINKED ISSUE #9250: [powermax] Expose panel status messages, and/or add a zone status to each Zone Thing
(Copied from https://community.openhab.org/t/powermax-binding/2824/289)

I see that you’re triggering the alarm_active channel when you get an 0xA5/0x04 message with bit 7 set in the system state flags (sysFlags in PowermaxStatusMessage.java). However, I’ve found that this only represents a _potential_ alarm condition, not necessarily a real one.

For example: if your system is armed and you violate a Delay zone (e.g. open your front door), you’ll immediately get a bit 7 message. If you walk to the keypad and disarm within the Delay period then there’s no alarm condition. If you don’t, then you have a real alarm. But the binding isn’t raising any event that correlates with this. If you want to write a rule that triggers on “my alarm is currently ringing,” there’s no way to do that.

A much better state for this is in the 0xA7 panel status message. You’re already decoding these in PowermaxPanelMessage.java but not really doing anything with them. In the example above, where you violate a Delay zone and then fail to disarm, you’ll get an 0xA7 message with an 0x03 (Delay Alarm) event at the instant the alarm starts ringing. Then it sends one of the Restore messages when it stops ringing. (0x1C General Restore if you disarm, and I’m not sure which one if you let it ring until the timeout.)

In any case… the 0xA7 messages are really useful. It would be great to have a channel that exposes these so you can trigger on various events. The system_status channel is useful but not detailed enough, and it’s a comma-separated list which you have to parse if you want to use it in a rule.

I’m also thinking about adding a Status channel to the Zone thing so you can more easily monitor and trigger on updates by zone. That channel could update on either an 0xA5 or 0xA7 message that affects the zone.

### PR DESCRIPTION
**Pull request by ronisaacson**
Closes #9250 .

* New panel channels:
    - Ringing indicator (the siren is currently sounding)
    - Date/time the last message was received from the panel
    - List of all active alarms and alerts (similar to panel's
      Memory list, but items get removed from the list as conditions
      resolve)
* New zone channels:
    - Alarmed indicator (zone is in alarm, or has an alarm in memory)
    - Tamper alarm indicator (same but for a tamper condition)
    - Inactive indicator
    - Tamper indicator (zone is actively tampered right now)
    - Last status message received for this zone, and when
* Use descriptive names for zones in log messages. If you create a
  Thing for a zone, it will use that Thing's label in all reporting
  for that zone. If there's no Thing then it will attempt to use the
  zone label from the panel (e.g. "Basement").
* Clear all channels during startup to keep from displaying stale
  values loaded from persistence
* Also includes some minor SAT fixes (checkstyle, spotbugs)

Signed-off-by: Ron Isaacson <isaacson.ron@gmail.com>

### CODE REVIEW

**Comment by ronisaacson** [comment]
Note: the CI build for this binding was [successful](https://ci.openhab.org/job/PR-openHAB-Addons/3977/org.openhab.addons.bundles$org.openhab.binding.powermax/). Build 3977 failed due to an unrelated error in a different binding.

**Comment by lolodomo** [comment]
>     * Clear all channels during startup to keep from displaying stale
>       values loaded from persistence

This looks like a bad setup of your persistence stuff. You are not forced to persist your items and you are not forced to setup "restoreOnStaryp" on all your items.
There is no reason for a binding to implement something because a supposed usage of a persistence service or even a bad setup of the persistence service.
By the way, I will see which channels are concerned by this change and if it really makes sense.

**Comment by ronisaacson** [comment]
> This looks like a bad setup of your persistence stuff. You are not forced to persist your items and you are not forced to setup "restoreOnStaryp" on all your items.

I'm open to suggestions on this. Right now it just uses the defaults, which I think persist all non-string items in rrd4j (more or less). I didn't do anything related to persistence but I notice that when OH restarts or the binding reloads or reconnects, all of the previous values come back.

If there's a way to disable persistence for some or all channels, that would probably be fine too.

**Comment by ronisaacson** [comment]
Hey @lolodomo, just want to see if you have any additional feedback. I've been running this build for a few months now and I'm happy with it. I have it send me a text message whenever the `active_alerts` channel changes. That one channel is enough to get detailed info on all alerts (trouble conditions) and alarms.

Is there anything else you want to see to move forward with this?

**Comment by lolodomo** [comment]
Needs to find more courage and time to test. M'y review has to be very serious as we are on a very sensitive (security) binding. But I will do, I have just forgotten.

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml [hunk 1]
```diff
@@ -17,6 +17,20 @@
 		</state>
 	</channel-type>
 
+	<channel-type id="last_message">
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml:20]: I would prefer last_message_time

Could be rather an advanced channel ?

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml [hunk 2]
```diff
@@ -17,6 +17,20 @@
...
+		<item-type>DateTime</item-type>
+		<label>System Last Message Received</label>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml:22]: Last Message Time

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml [hunk 3]
```diff
@@ -31,6 +45,13 @@
 		<state readOnly="true"></state>
 	</channel-type>
 
+	<channel-type id="ringing" advanced="true">
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml:48]: Could rather be not advanced ?

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml [hunk 4]
```diff
@@ -121,6 +172,20 @@
 		<state readOnly="true"></state>
 	</channel-type>
 
+	<channel-type id="zone_last_message">
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml:175]: I suggest last_status

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml [hunk 5]
```diff
@@ -121,6 +172,20 @@
...
+		<item-type>String</item-type>
+		<label>Zone Last Status Message</label>
+		<description>The most recent status message reported by the zone</description>
+		<state readOnly="true" pattern="%s"></state>
+	</channel-type>
+
+	<channel-type id="zone_last_message_time">
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/channels.xml:182]: I suggest last_status_time

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/ip.xml [hunk 6]
```diff
@@ -17,8 +17,11 @@
 			<channel id="with_zones_bypassed" typeId="with_zones_bypassed"/>
 			<channel id="trouble" typeId="trouble"/>
 			<channel id="alert_in_memory" typeId="alert_in_memory"/>
+			<channel id="ringing" typeId="ringing"/>
 			<channel id="pgm_status" typeId="pgm_status"/>
 			<channel id="mode" typeId="mode"/>
+			<channel id="last_message" typeId="last_message"/>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/ip.xml:23]: To be coherent with my suggestion: `<channel id="last_message_time" typeId="last_message_time"/>`

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/serial.xml [hunk 7]
```diff
@@ -17,8 +17,11 @@
 			<channel id="with_zones_bypassed" typeId="with_zones_bypassed"/>
 			<channel id="trouble" typeId="trouble"/>
 			<channel id="alert_in_memory" typeId="alert_in_memory"/>
+			<channel id="ringing" typeId="ringing"/>
 			<channel id="pgm_status" typeId="pgm_status"/>
 			<channel id="mode" typeId="mode"/>
+			<channel id="last_message" typeId="last_message"/>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/serial.xml:23]: To be coherent with my suggestion: `<channel id="last_message_time" typeId="last_message_time"/>`

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/zone.xml [hunk 8]
```diff
@@ -21,6 +21,12 @@
 			<channel id="last_trip" typeId="last_trip"/>
 			<channel id="low_battery" typeId="system.low-battery"/>
 			<channel id="bypassed" typeId="bypassed"/>
+			<channel id="alarmed" typeId="alarmed"/>
+			<channel id="tamper_alarm" typeId="tamper_alarm"/>
+			<channel id="inactive" typeId="inactive"/>
+			<channel id="tampered" typeId="tampered"/>
+			<channel id="zone_last_message" typeId="zone_last_message"/>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/zone.xml:28]: To be coherent with my suggestion: `<channel id="last_status" typeId="last_status"/>`

File: bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/zone.xml [hunk 9]
```diff
@@ -21,6 +21,12 @@
...
+			<channel id="zone_last_message_time" typeId="zone_last_message_time"/>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/resources/OH-INF/thing/zone.xml:29]: To be coherent with my suggestion: `<channel id="last_status_time" typeId="last_status_time"/>`

File: bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowermaxZoneState.java [hunk 10]
```diff
@@ -28,7 +28,13 @@
     public DateTimeValue lastTripped = new DateTimeValue(this, LAST_TRIP);
     public BooleanValue lowBattery = new BooleanValue(this, LOW_BATTERY);
     public BooleanValue bypassed = new BooleanValue(this, BYPASSED);
+    public BooleanValue alarmed = new BooleanValue(this, ALARMED);
+    public BooleanValue tamperAlarm = new BooleanValue(this, TAMPER_ALARM);
+    public BooleanValue inactive = new BooleanValue(this, INACTIVE);
+    public BooleanValue tampered = new BooleanValue(this, TAMPERED);
     public BooleanValue armed = new BooleanValue(this, ARMED);
+    public StringValue lastMessage = new StringValue(this, ZONE_LAST_MESSAGE);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowermaxZoneState.java:36]: To be coherent with my suggestion: LAST_STATUS

File: bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowermaxZoneState.java [hunk 11]
```diff
@@ -28,7 +28,13 @@
...
+    public DateTimeValue lastMessageTime = new DateTimeValue(this, ZONE_LAST_MESSAGE_TIME);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowermaxZoneState.java:37]: To be coherent with my suggestion: LAST_STATUS_TIME

File: bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowermaxBindingConstants.java [hunk 12]
```diff
@@ -58,8 +61,14 @@
     public static final String TRIPPED = "tripped";
     public static final String LAST_TRIP = "last_trip";
     public static final String BYPASSED = "bypassed";
+    public static final String ALARMED = "alarmed";
+    public static final String TAMPER_ALARM = "tamper_alarm";
+    public static final String INACTIVE = "inactive";
+    public static final String TAMPERED = "tampered";
     public static final String ARMED = "armed";
     public static final String LOCKED = "locked";
+    public static final String ZONE_LAST_MESSAGE = "zone_last_message";
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowermaxBindingConstants.java:70]: To be coherent with my suggestion: LAST_STATUS = "last_status"

File: bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowermaxBindingConstants.java [hunk 13]
```diff
@@ -58,8 +61,14 @@
...
+    public static final String ZONE_LAST_MESSAGE_TIME = "zone_last_message_time";
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowermaxBindingConstants.java:71]: To be coherent with my suggestion: LAST_STATUS_TIME = "last_status_time"

File: bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowermaxBindingConstants.java [hunk 14]
```diff
@@ -47,8 +47,11 @@
 
     // List of all Channel ids
     public static final String MODE = "mode";
+    public static final String LAST_MESSAGE = "last_message";
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowermaxBindingConstants.java:50]: To be coherent with my suggestion: LAST_MESSAGE_TIME = "last_message_time"

**Review by lolodomo — CHANGES_REQUESTED** [review]
Review part 1 : channels

File: bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowermaxState.java [hunk 15]
```diff
@@ -96,6 +128,36 @@ public PowermaxState(PowermaxPanelSettings panelSettings, TimeZoneProvider timeZ
...
+        activeAlertQueue = new ArrayList<>();
+    }
+
+    /**
+     * New PowermaxState objects are created with null values for most fields.
+     * When the binding starts up, to prevent fields from being shown as null
+     * or using persisted values which may be stale, we'll initialize a known
+     * empty state for many fields.
+     */
+    public void setInitialState() {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowermaxState.java:140]: To be suppressed

**Review by lolodomo — CHANGES_REQUESTED** [review]
Final part of my review.

**Comment by lolodomo** [comment]
@ronisaacson : my review is now finished and that looks very good in general. To summarize, I am suggesting new names for only few added channels and your workaround introduced in the binding to fix one problem you have with persistence is not welcome ;)
I will now test your changes, at least partially, as I don't really want to make my alarm ringing...

File: bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowermaxState.java [hunk 16]
```diff
@@ -214,6 +276,80 @@ public void updateZoneInfo(int zoneIdx, int zoneInfo) {
...
+    }
+
+    public String getActiveAlerts() {
+        if (activeAlertList == null) {
+            return null;
+        }
+
+        if (activeAlertList.isEmpty()) {
+            return "None";
+        }
```
Comment (lolodomo) [inline bundles/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowermaxState.java:297]: Please return "None" when `activeAlertList` is null too, it allows having the channel initialized to "None" after binding startup.

**Review by lolodomo — CHANGES_REQUESTED** [review]
Anoter request change after testing

**Comment by lolodomo** [comment]
Even after suppressing `setInitialState`, all channels are set as expected after binding startup except the "ringing" channel which is still not set. It would be better if it could be set to OFF after the init sequence.

**Comment by ronisaacson** [comment]
@lolodomo thanks for your patience. Now that the persistence discussion is wrapped up, let's get this finished. If you have a branch with some or all of your changes, yes I'll happily take a PR. I'll look at the initialization of Ringing as well.

**Comment by lolodomo** [comment]
I can prepare one during this week.

**Comment by ronisaacson** [comment]
No don't go to any trouble, I'll start working on the changes shortly.

**Comment by lolodomo** [comment]
I checked again what channels are not initialized (with my updated version opf the PR) after a full init sequence (reading of panel setup):

- new channel "ringing": my feeling is that this one should be set properly
- event_log_X but this is expected (will be set only when requesting the events history)
- update-event_logs and download_setup but this is expected as they are not channels with a real state
- last_trip but this is normal (will be set the next tiume the zone is tripped)
- new channels last_message and mast_message_time but this is normal (will be set when a new event will be received)

So except for the new channel "ringing", everything looks good to me.

Note that in "standard" mode (not powerlink mode), maybe more channels are not initialized, I have not tested it. Maybe something to test, especially for the new channels introduced by this PR.

**Comment by lolodomo** [comment]
Just tested with the "standard" mode and good news, it makes no difference regarding channels initialization.

**Review by lolodomo — APPROVED** [review]
LGTM

**Comment by lolodomo** [comment]
Thank you

**Pull request merged**